tfhe/src/core_crypto/fft_128_split_impl/crypto/bootstrap.rs

```diff
@@ -0,0 +1,236 @@
+                )
+            }
+
+            // We initialize the ct_0 used for the successive cmuxes
+            let mut ct0_lo = lut_lo;
+            let mut ct0_hi = lut_hi;
+
+            for (lwe_mask_element, bootstrap_key_ggsw) in
+                izip!(lwe_mask.iter(), this.into_ggsw_iter())
+            {
+                if *lwe_mask_element != 0 {
+                    let stack = stack.rb_mut();
+                    // We copy ct_0 to ct_1
+                    let (mut ct1_lo, stack) =
+                        stack.collect_aligned(CACHELINE_ALIGN, ct0_lo.as_ref().iter().copied());
+                    let (mut ct1_hi, stack) =
+                        stack.collect_aligned(CACHELINE_ALIGN, ct0_hi.as_ref().iter().copied());
+                    let mut ct1_lo = GlweCiphertextMutView::from_container(
+                        &mut *ct1_lo,
+                        ct0_lo.polynomial_size(),
+                    );
+                    let mut ct1_hi = GlweCiphertextMutView::from_container(
+                        &mut *ct1_hi,
+                        ct0_lo.polynomial_size(),
+                    );
+
+                    // We rotate ct_1 by performing ct_1 <- ct_1 * X^{a_hat}
+                    for (poly_lo, poly_hi) in izip!(
+                        ct1_lo.as_mut_polynomial_list().iter_mut(),
+                        ct1_hi.as_mut_polynomial_list().iter_mut(),
+                    ) {
+                        polynomial_wrapping_monic_monomial_mul_assign_split(
+                            poly_lo,
+                            poly_hi,
+                            MonomialDegree(pbs_modulus_switch(
+                                *lwe_mask_element,
+                                lut_poly_size,
+                                ModulusSwitchOffset(0),
+                                LutCountLog(0),
+                            )),
+                        )
+                    }
+
+                    // ct1 is re-created each loop it can be moved, ct0 is already a view, but
+                    // as_mut_view is required to keep borrow rules consistent
+                    cmux_split(
+                        &mut ct0_lo,
+                        &mut ct0_hi,
+                        &mut ct1_lo,
+                        &mut ct1_hi,
+                        &bootstrap_key_ggsw,
+                        fft,
+                        stack,
+                    );
+                }
+            }
+        }
+        implementation(
+            self.as_view(),
+            lut_lo.as_mut_view(),
+            lut_hi.as_mut_view(),
+            lwe.as_view(),
+            fft,
+            stack,
+        )
+    }
+
+    pub fn bootstrap_u128<ContLweOut, ContLweIn, ContAcc>(
+        &self,
+        lwe_out: &mut LweCiphertext<ContLweOut>,
+        lwe_in: &LweCiphertext<ContLweIn>,
+        accumulator: &GlweCiphertext<ContAcc>,
+        fft: Fft128View<'_>,
+        stack: PodStack<'_>,
+    ) where
```

Comment:
I think we can have a "Bootstrap" trait that is implemented for that function allowing to easily test the various implementations and can allow to abstract over it if needed in higher level APIs

Hear me out:
LweOut is always an LweCiphertext with same Scalar type as LweIn (also an LweCiphertext) and as the accumulator a GlweCiphertext

stack is always a PodStack

The only thing that may be slightly annoying is managing the Fft, but we could have a sealed FftView trait implement for FftView and FftView128, then the trait implementation is basically just calling the underlying implementation of each BSK, what do you think ?

We can even force the Fft struct we expect with an associated type to make it less error prone, and we can have a parameterization over the Scalar, worst case we just call the methods directly in the higher levels without using the trait, but at least the common trait allows to easily test Bootstrap version on the same high level code